File: dynabuffers-java/src/main/kotlin/dynabuffers/ast/datatype/RefType.kt
package dynabuffers.ast.datatype

import dynabuffers.api.IRegistry
import dynabuffers.api.ISerializable
import dynabuffers.api.IType
import java.nio.ByteBuffer

class RefType(val options: RefTypeOptions) : IType, ISerializable {

    override fun size(value: Any, registry: IRegistry) = registry.resolve(options.name).size(value, registry)

    override fun serialize(value: Any, buffer: ByteBuffer, registry: IRegistry) {
        registry.resolve(options.name).serialize(value, buffer, registry)
    }

    override fun deserialize(buffer: ByteBuffer, registry: IRegistry): Any {
        return registry.resolve(options.name).deserialize(buffer, registry)
    }

    override fun supports(value: Any): Boolean {
        return value is Map<*, *>
    }

    data class RefTypeOptions(val name: String)

}

File: dynabuffers-python/dynabuffers/ast/datatype/RefType.py
from dynabuffers.api.ISerializable import ISerializable


class RefTypeOptions:
    def __init__(self, name: str):
        self.name = name


class RefType(ISerializable):
    def __init__(self, options: RefTypeOptions):
        self.options = options

    def size(self, value, registry):
        return registry.resolve(self.options.name).size(value, registry)

    def serialize(self, value, buffer, registry):
        registry.resolve(self.options.name).serialize(value, buffer, registry)

    def deserialize(self, buffer, registry):
        return registry.resolve(self.options.name).deserialize(buffer, registry)
